streaming/base/dataset/streaming.py

```diff
@@ -0,0 +1,623 @@
+# Copyright 2022 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""A streaming pytorch IterableDataset, resumable mid-epoch, whose shards reside locally."""
+
+import json
+import os
+from enum import IntEnum
+from multiprocessing.shared_memory import SharedMemory
+from threading import Thread
+from time import sleep
+from typing import Any, Dict, Iterator, Optional, Tuple
+
+import numpy as np
+import torch
+from filelock import FileLock
+from numpy.typing import NDArray
+from torch.utils.data import IterableDataset
+
+from streaming.base import distributed as dist
+from streaming.base.compression import decompress
+from streaming.base.download import download
+from streaming.base.format import reader_from_json
+from streaming.base.format.base.reader import FileInfo
+from streaming.base.hashing import get_hash
+from streaming.base.index import Index, get_index_basename
+from streaming.base.shared import SharedBarrier
+from streaming.base.shuffle import get_shuffle
+from streaming.base.world import World
+
+# Time to wait, in seconds.
+TICK = 0.07
+
+
+class _ShardState(IntEnum):
+    """The download status of a shard.
+
+    Restrictions:
+    - The initial state of UNKNOWN must be zero.
+    - The state will only ever change in the upward direction.
+    """
+    UNKNOWN = 0
+    DOWNLOADING = 1
+    DOWNLOADED = 2
+
+
+class _PartitionState:
+    """The download status of a partition of samples.
+
+    Args:
+        sample_ids (NDArray[np.int64]): This worker's partition of the sample space.
+    """
+
+    def __init__(self, sample_ids: NDArray[np.int64]) -> None:
+        self.sample_ids = sample_ids
+        self.total = len(sample_ids)
+        self.iter_index = 0
+        self.download_index = 0
+        self.is_stopped = False
+
+    def stop(self) -> None:
+        """Stop the thread and exit."""
+        self.is_stopped = True
+
+    def __iter__(self) -> Iterator[int]:
+        """Iterate over our samples while waiting for them to download first.
+
+        Returns:
+            Iterator[int]: Each sample, having been downloaded.
+        """
+        while self.iter_index < self.total:
+            if self.iter_index < self.download_index:
+                yield self.sample_ids[self.iter_index]
+                self.iter_index += 1
+                continue
+            if self.is_stopped:
+                break
+            sleep(TICK)
+
+
+class Dataset(IterableDataset):
```

Comment:
Do we want to rename to `StreamingDataset` based on poll in slack?

Reply:
renamed